### DISCUSSION TITLE
Tests suite fails for the file-search package

### DISCUSSION THREAD

**Discussion by akhileshraju** [eclipse-theia/theia#11386] [Q&A]
I cloned the main branch, ran `yarn` at the top level directory followed by `yarn test` to run all the tests. I see that the test for `@theia/file-search` fails with the following error. This isn't expected right?

<img width="804" alt="image" src="https://user-images.githubusercontent.com/17533956/177050717-9a05d527-a14b-4946-b63d-a8c0c48a54e9.png">

I am using Node `v16.15.1` on macOS `12.4`

**Comment by vince-fugnitto** [comment]
@akhileshraju thank you for the discussion, it should definitely not be failing. I've tried locally and it passes, as well as during [CI](https://github.com/eclipse-theia/theia/runs/7139063402?check_suite_focus=true#step:6:1381). You may want to check if you are missing some [prerequisite](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#prerequisites) if you haven't already done so.

_Steps_:

- `yarn`
- `yarn download:plugins` (download the bundled vscode extensions which are later tested against)
- `yarn browser build` (build the example-browser application which is later tested against)
- `yarn test`

_Additional Info_:

- macOS `12.3`
- Node: `v16.14.0`

**Reply by akhileshraju** [reply]
Hmm I tried those steps, still no change for me.
Could you shed some light on what that error might mean? Maybe that may give us some clue about what might be the cause

**Reply by akhileshraju** [reply]
I do have all the mentioned pre-requisites

**Reply by vince-fugnitto** [reply]
The error itself is not telling me much, do you have any additional info?
It’s also odd that locally and during CI it passes fine.

**Reply by akhileshraju** [reply]
What other kind of info can I provide to help with debug?

I also tried below steps of 2 other platforms

- Windows 10
- Ubuntu 18.04

> Steps:
>
> yarn
> yarn download:plugins (download the bundled vscode extensions which are later tested against)
> yarn browser build (build the example-browser application which is later tested against)
> yarn test

I see the `@theia/file-search` fails again but the `ASSERTION_ERR` is for a different file

-  on macOS it was for `.vscode/theia.code-snippets` 
- on Windows it's for `.github/CODEOWNERS` 
- on Ubuntu it's for `CODE_OF_CONDUCT.md`

**Ubuntu**

![image](https://user-images.githubusercontent.com/17533956/177609958-49d2cbb5-7d0f-493c-bfd3-ebf7f05b8bb6.png)

**Windows 10**

![image](https://user-images.githubusercontent.com/17533956/177610140-648d74cc-86d0-438f-9209-92647f15581c.png)


All 3 platforms have identical Node `v16.15.1`

**Reply by colin-grant-work** [reply]
@akhileshraju, it looks like the test that is failing is this one:

https://github.com/eclipse-theia/theia/blob/fb5d6955a610ef22487454b612370c2454d518b2/packages/file-search/src/node/file-search-service-impl.spec.ts#L166-L175

If you go into that file and change the code for the test to this:

```typescript
        it('not fuzzy', async () => {
            const searchPattern = rootUri.path.dir.base;
            const matches = await service.find(searchPattern, { rootUris: [rootUri.toString()], fuzzyMatch: false, useGitIgnore: true, limit: 200 });
            for (const match of matches) {
                const relativeUri = rootUri.relative(new URI(match));
                assert.notStrictEqual(relativeUri, undefined);
                const relativeMatch = relativeUri!.toString();
                assert(relativeMatch.includes(searchPattern), `${relativeMatch} should include ${searchPattern}`);
            }
        });
```

it may provide more helpful output than the current test and give a pointer to how we can fix the test to pass in your environment.

**Reply by akhileshraju** [reply]
Thanks for that pointer @colin-grant-work.

Here's what it prints now

```

  16 passing (415ms)
  1 pending
  1 failing

  1) search-service
       irrelevant absolute results
         not fuzzy:

      AssertionError [ERR_ASSERTION]: CODE_OF_CONDUCT.md should include code
      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (lib/node/file-search-service-impl.spec.js:148:17)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)



Exit with failure status (1): nyc mocha --opts ../../configs/mocha.opts "./lib/**/*.*spec.js"
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by vince-fugnitto** [reply]
I investigated a bit and it seems as though this test (which has been written a while ago) is poorly written, it is dependent on where `theia` is cloned and uses that as a `searchPattern`. Instead we should update it to make it more robust, and even use [test-resources](https://github.com/eclipse-theia/theia/tree/master/packages/file-search/test-resources) as the directory to search in.

We can see that the `searchPattern` should not be determined in such a way:

```
result: { pattern: vince, matches len: 0, root: file:///Users/test/vince/theia }
```

---

Can you try the following?

```diff
it('not fuzzy', async () => {
+   const searchPattern = 'package'; // package.json should produce a result.
-   const searchPattern = rootUri.path.dir.base;
    const matches = await service.find(searchPattern, { rootUris: [rootUri.toString()], fuzzyMatch: false, useGitIgnore: true, limit: 200 });
+   expect(matches).not.empty;
    for (const match of matches) {
        const relativeUri = rootUri.relative(new URI(match));
        assert.notStrictEqual(relativeUri, undefined);
        const relativeMatch = relativeUri!.toString();
        assert.notStrictEqual(relativeMatch.indexOf(searchPattern), -1, relativeMatch);
    }
});
```

**Reply by akhileshraju** [reply]
With the above change, that test passes but something way down fails

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/17533956/178124413-15858b75-1480-4f6a-9986-ebe48f4c490d.png">

**Reply by vince-fugnitto** [reply]
> With the above change, that test passes but something way down fails
> 
> <img alt="image" width="1002" src="https://user-images.githubusercontent.com/17533956/178124413-15858b75-1480-4f6a-9986-ebe48f4c490d.png">

The api tests (integration) can sometimes be temperamental. I for instance never run them locally but only do so during CI where I re-run the test if it fails.
